crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -53,17 +54,34 @@ impl LoopState {
     }
 }
 
+/// Used in the scope-stack of the [`SemanticIndexBuilder`] to keep track of whether
+/// or not the current scope belongs to a method in a class, which is the case if
+/// the current scope is a `BuilderScopeKind::FunctionBody` and the parent scope is
+/// a `BuilderScopeKind::ClassBody`.
+///
+/// This is called `BuilderScopeKind` to distinguish it from the `ScopeKind` enum,
+/// which is not fine-grained enough for this use case (functions and lambdas are
+/// both `ScopeKind::Function`)
+#[derive(Clone, Copy, Debug, PartialEq)]
+enum BuilderScopeKind {
+    ClassBody,
+    FunctionBody,
+    Other,
+}
+
 pub(super) struct SemanticIndexBuilder<'db> {
     // Builder state
     db: &'db dyn Db,
     file: File,
     module: &'db ParsedModule,
-    scope_stack: Vec<(FileScopeId, LoopState)>,
+    scope_stack: Vec<(FileScopeId, LoopState, BuilderScopeKind)>,
```

Comment:
nit: is it time to create a `ScopeInfo` struct so that we can access named fields on the items in this stack?

Reply:
My suggestion is that we may not need to add a third item to this stack in this PR? But making it a struct could be a good idea regardless.